### PR TITLE
Modification of Polygon::convexHull() method

### LINKED ISSUE #34: Performance of convexHull()
Here is the screen of recently added performance tests for Polygon::convexHull()
image
tests were run with Release build on the machine: 3.1GHz, 4-core CPU.

As you may see convexHull computation took in average 22ms for polygon with 10000 points and 237ms for 100000 points polygon. Looks like current convexHull algorithm is working correct and fast enough. Nevertheless, this is one of the key function the whole project, so I'm wondering if we may still speed it up.

**My suggestion:**
Please review this monotone chain algorithm for convex hull: https://en.wikibooks.org/wiki/Algorithm_Implementation/Geometry/Convex_hull/Monotone_chain.
Try to implement this alternative algorithm in our project. It would not take much time, since there is already C++ implementation present, just adapt it to our code. Compare performance of current convexHull implementation and this alternative version. If you will see notable divergence in performance, please try to figure out and explain why.
**Note:** this is not a bug, but consider this task as a research and possible improving work. It's not high-priority, but still it should be done.

### PR DESCRIPTION
**Pull request by IlliaHryshchuk**
After implementation of new algorithm the speed of workflow increased in 50-60 times:

image
image
image
image

I think that so giant increasing occurs because we mustn't use `Polygon::orientation()` method, and all points are lexicographically sorted before the algorithm.
Also in my previous algorithm were 3 nested loops : 
do{
    for()
    {   
          for(){}//inside Polygon::orientation()
    }   
}while();

and it cause O(n^3) complexity, while the new algorithm has 3 consecutive cycles that is 3O(n) or just O(n).

### CODE REVIEW

File: GeometryCore/src/Polygon.cpp [hunk 1]
```diff
@@ -130,24 +130,30 @@ namespace geom_utils
     Polygon Polygon::convexHull() const 
     { 
         if (points.size() <= 3) return *this; //can't bild a polygon from 2 points or less
+
+        auto my_cross = [](const FPoint2D& O, const FPoint2D& A, const FPoint2D& B) {return (A.x - O.x) * (B.y - O.y) - (A.y - O.y) * (B.x - O.x); };
```
Comment (VladGribinchuk) [inline GeometryCore/src/Polygon.cpp]: Please dont re-write things twice. We have already implemented vector cross product function - please see LinearAlgebra.

**Review by VladGribinchuk — CHANGES_REQUESTED** [review]

**Comment by VladGribinchuk** [comment]
Try to investigate a bit and let me know your thoughts about that difference in performance. Why new method is faster.

**Comment by IlliaHryshchuk** [comment]
> Try to investigate a bit and let me know your thoughts about that difference in performance. Why new method is faster.

Just was writting it. I added my understanding to the description of pull request

File: GeometryCore/src/Polygon.cpp [hunk 2]
```diff
@@ -144,13 +142,13 @@ namespace geom_utils
 
         // Build lower hull
         for (size_t i = 0; i < pointsCopy.size(); ++i) {
-            while (k >= 2 && my_cross(convexHull.points[k - 2], convexHull.points[k - 1], pointsCopy[i]) <= 0) k--;
+            while (k >= 2 && geom_utils::cross(FPoint2D(convexHull.points[k - 2] - pointsCopy[i]), FPoint2D(convexHull.points[k - 1] - pointsCopy[i])) <= 0) k--;
```
Comment (VladGribinchuk) [inline GeometryCore/src/Polygon.cpp]: Please check carefully if this new version is identical to the previous cross(OA, OB) vectors... at the first glance, looks like you have wrong cross product

**Comment by VladGribinchuk** [comment]
Good, your description of performance difference looks trustworthy, you are right about complexity. 

But also please take note that previous convexHull version contains orientation() calls, like you wrote, which include parallel computation `#pragma omp parallel for` inside and are being called multiple times in the loop. Please investigate how this impact on performance (hint: try with enabled/disabled this omp parallel). Let me please know the results and your conclusion.

**Comment by IlliaHryshchuk** [comment]
without openmp it runs 2 times faster than with it. So it means that the program spend more time for creation of threads, and doesnt get profit from it. It mean that it is better to delete openmp from  `Polygon::area()` ?

image

File: GeometryCore/src/Polygon.cpp (hunk 2 above)
Comment (IlliaHryshchuk) [reply GeometryCore/src/Polygon.cpp]: Yes, it is my mistake. I will fix it

**Comment by VladGribinchuk** [comment]
> without openmp it runs 2 times faster than with it. So it means that the program spend more time for creation of threads, and doesnt get profit from it. It mean that it is better to delete openmp from Polygon::area() ?

Your conclusion is correct... by a half. You're right that creation of threads takes some time, **and this should be taken into account while developing**. But it doesnt mean that program doesnt get profit from it. All that matters is context. As you see, here we have numerous calls from the loop. As result, we got a situation with numerous creation/destroying of resources needed for parallel threading which takes large amount of time. Though the area() method itself is good optimized by openMP.

**Comment by IlliaHryshchuk** [comment]
Ok, i have understood it, so I can commit my cross-fix

**Review by VladGribinchuk — APPROVED** [review]
Looks good now

**Pull request merged**